csharp/ql/src/semmle/code/csharp/frameworks/EntityFramework.qll

```diff
@@ -0,0 +1,192 @@
+      c = this.getAConstructor() and
+      source.(CallableFlowSourceArg).getArgumentIndex() = 0 and
+      sink instanceof CallableFlowSinkReturn and
+      preservesValue = true
+      or
+      c = this.getAConversionTo() and
+      source.(CallableFlowSourceArg).getArgumentIndex() = 0 and
+      sink instanceof CallableFlowSinkReturn and
+      preservesValue = true
+    }
+
+    ConversionOperator getAConversionTo() {
+      result = this.getAMember() and
+      result.getTargetType() instanceof RawSqlStringStruct and
+      result.getSourceType() instanceof StringType
+    }
+  }
+
+  /**
+   * A parameter that accepts raw SQL. Parameters of type `System.FormattableString`
+   * are not included as they are not vulnerable to SQL injection.
+   */
+  private class SqlParameter extends Parameter {
+    SqlParameter() {
+      this.getType() instanceof StringType and
+      (
+        exists(Callable c | this = c.getParameter(0) | c.getName().matches("%Sql"))
+        or
+        this.getName() = "sql"
+      ) and
+      this.getCallable().getDeclaringType().getDeclaringNamespace().getParentNamespace*() instanceof
+        EFNamespace
+      or
+      this.getType() instanceof RawSqlStringStruct
+      or
+      this = any(RawSqlStringStruct s).getAConstructor().getAParameter()
+      or
+      this = any(RawSqlStringStruct s).getAConversionTo().getAParameter()
+    }
+  }
+
+  /** A call to a method in EntityFrameworkCore that executes SQL. */
+  class EntityFrameworkCoreSqlSink extends SqlExpr, Call {
+    SqlParameter sqlParam;
+
+    EntityFrameworkCoreSqlSink() { this.getTarget().getAParameter() = sqlParam }
+
+    override Expr getSql() { result = this.getArgumentForParameter(sqlParam) }
+  }
+
+  /** A call to `System.Data.Entity.DbSet.SqlQuery`. */
+  class SystemDataEntityDbSetSqlExpr extends SqlExpr, MethodCall {
+    SystemDataEntityDbSetSqlExpr() {
+      this.getTarget() = any(SystemDataEntity::DbSet dbSet).getSqlQueryMethod()
+    }
+
+    override Expr getSql() { result = this.getArgumentForName("sql") }
+  }
+
+  /** A call to a method in `System.Data.Entity.Database` that executes SQL. */
+  class SystemDataEntityDatabaseSqlExpr extends SqlExpr, MethodCall {
+    SystemDataEntityDatabaseSqlExpr() {
+      exists(SystemDataEntity::Database db |
+        this.getTarget() = db.getSqlQueryMethod() or
+        this.getTarget() = db.getExecuteSqlCommandMethod() or
+        this.getTarget() = db.getExecuteSqlCommandAsyncMethod()
+      )
+    }
+
+    override Expr getSql() { result = this.getArgumentForName("sql") }
+  }
+
+  /**
+   * A dataflow node whereby data flows from a property write to a property read
+   * via some database. The assumption is that all writes can flow to all reads.
+   */
+  class MappedPropertyJumpNode extends DataFlow::NonLocalJumpNode {
+    MappedProperty property;
+
+    MappedPropertyJumpNode() { this.asExpr() = property.getAnAssignedValue() }
+
+    override DataFlow::Node getAJumpSuccessor() {
```

Comment:
I think this should perhaps be a taint step, although I realize that technically it preserves the value.